pyro/poutine/runtime.py

```diff
@@ -196,3 +151,52 @@ def am_i_wrapped():
     :returns: bool
     """
     return len(_PYRO_STACK) > 0
+
+
+def effectful(fn=None, type=None):
+    """
+    Wrapper for calling apply_stack to apply any active effects.
+    """
+    if fn is None:
+        return lambda x: effectful(x, type=type)
+
+    if getattr(fn, "__wrapped", None):
+        return fn
+    else:
+        if type is None:
+            type = fn.__code__.co_name
+
+        assert type != "message", "cannot use 'message' as keyword"
+
+        def _fn(*args, **kwargs):
```

Comment:
Do you know how expensive it is to create a closure each sample statement? I do not know, but I'm a little worried about our gradual increase of Python overhead over time. It is nice to find zero-abstraction-cost alternatives whenever possible.

Reply:
Inner function overhead is discussed a little [here](https://stackoverflow.com/questions/7839632/is-there-an-overhead-when-nesting-functions-in-python), not sure about closures but I guess the cost is one extra variable lookup (a `LOAD_FAST` VM op) each for `fn` and `type` relative to the existing situation for `sample`.

I think most of our Python overhead comes from the dispatching (`getattr`) in `poutine.runtime.apply_stack`, and this is unlikely to be large in comparison after `effectful` is memoized.  I'll do some profiling when I get a chance.

Reply:
I've reverted the definition of `pyro.sample`, so that performance should be affected.  In the other use cases I imagine for this, `effectful` is only called once per function/operation definition, rather than once per function call.